### PR TITLE
Mix the productPrefix.hashCode into the hash code of case classes

### PR DESCRIPTION
**Pull request by adriaanm**
Fixes scala/bug#10495

### CODE REVIEW

**Comment by adriaanm** [comment]
Test failures mostly look like a matter of check file updates, but this one is kind of puzzling:

```diff
!![0m  870 - run/t2873.scala                           [output differs]
% [0mdiff /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/t2873-run.log /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/t2873.check
@@ -1 +1 @@
-RedBlack<A>$Empty$
+RedBlack<A>.Empty$

##### Log file '/home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/t2873-run.log' from failed test #####
[0m[0m
RedBlack<A>$Empty$
```

any ideas?

**Review by Ichoran — APPROVED** [review]
The core change to MurmurHash3 looks good to me.  The explicit syntax tree building isn't something I can check for correctness by eye, but it looks plausibly correct.  I don't know what the source of the errors is, and I don't have time to check.  It doesn't look like it's anything wrong with the hashing itself, though.

(I'm "approving" in principle; obviously the test failures have to be fixed somehow.)

File: src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala [hunk 1]
```diff
@@ -276,6 +276,11 @@ trait SyntheticMethods extends ast.TreeDSL {
       createMethod(nme.hashCode_, Nil, IntTpe) { m =>
         val accumulator = m.newVariable(newTermName("acc"), m.pos, SYNTHETIC) setInfo IntTpe
         val valdef      = ValDef(accumulator, Literal(Constant(0xcafebabe)))
+        val mixPrefix   =
+          Assign(
+            Ident(accumulator),
+            callStaticsMethod("mix")(Ident(accumulator),
+              Apply(gen.mkAttributedSelect(gen.mkAttributedSelect(mkThis, Product_productPrefix), Object_hashCode), Nil)))
```
Comment (Ichoran) [inline src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala:283]: I don't understand why `Nil` is here, but maybe that means "use the default value; it wasn't supplied"?

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala:283]: it's putting the `()` in `acc = this.productPrefix.hashCode()`

**Comment by adriaanm** [comment]
Hmmm.. This does violate the [Object#hashCode contract](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode--):

> If two objects are equal according to the equals(Object) method, then calling the hashCode method on each of the two objects must produce the same integer result.

EDIT: it doesn't -- my mistake

**Comment by adriaanm** [comment]
resubmitting

**Pull request closed without merging**